lib/zabbix/zabbix.go

```diff
@@ -35,12 +35,44 @@ type Result struct {
 // Results struct holds a list of Zabbix host info
 type Results []Result
 
-// NewAPI builds new API struct
-func NewAPI(URL, key string) API {
-	return API{
+// API response Error struct
+type apiError struct {
+	Message string `json:"message,omitempty"`
+	Data    string `json:"data,omitempty"`
+}
+
+// CreateClientUsingAPIKey creates new API client using key
+func CreateClientUsingAPIKey(URL, key string) (*API, error) {
+	if URL == "" {
+		return nil, fmt.Errorf("zabbix URL missing. please run setup again")
+	}
+
+	if key == "" {
+		return nil, fmt.Errorf("zabbix API key missing. please run setup again")
+	}
+
+	return &API{
```

Comment:
If you do this, you will need to expose an interface instead 
e.g.
```go
type API interface {
    GetKey() (string, error)
    GetHostsInfo() (http.Response, error)
    BuildPayload() (*payloadtype*, error)  
    // ...etc
}
```

Then hint that the interface type is returned in the api methods _e.g._

```go
func CreateClientUsingAuth(URL, user, password string) (API, error) {}
```

Exposing an interface instead of the structure/implementation other than allowing for users easily write an implementation of the API in their tests also haves you consider what methods you really what to be public.

Reply:
Hmm ... returning interfaces is a big no-no in goland. The idiom is _[accept](https://mycodesmells.com/post/accept-interfaces-return-struct-in-go) [interfaces](https://www.reddit.com/r/golang/comments/5cos3x/accept_interfaces_return_structs/) [return concrete types(structs)](https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8)._

In any case, @osule I'm curious as to why going to great length to "hide API" 🤔 It seems to be the central part of the domain logic of the library and every other thing are, merely helpers; utilities to make working with it easier. If that's correct, then it might make sense to not restrict it's flexibility to the flavors provided by the utilities. An example from the standard library is the [http#Response](https://godoc.org/net/http#Response). While there are helper functions to build different flavors of it, you can also construct (and work with) it's base form. The burden of the domain is on the `response` object and not how it was constructed.

Reply:
Generally, I'll favor not having the user construct the api struct (it minimizes errors from constructing it the wrong way)

On returning concrete types over interfaces, I'll say it depends on the context. A number of standard packages do differently. 
For one looking at the [io package](https://golang.org/pkg/io/#LimitReader), LimitReader returns a Reader. I like to guess that the intent there is to allow for the caller to be able to pass and receive an implementation of the interface.
I share the concern that it's going too far to hide the API. I don't currently see a reason for callers to have a different implementation for API and no better way to keep the user from constructing the struct the wrong way. 

Perhaps, I'm misjudging the outcome of this project. This is going to be used as a command line tool and not a library AFAICT, so it's fine to keep the API struct as is.

Reply:
okay cool. Is this an approval of the PR or not yet? 🤔